Support Objects overhaul. List should be comprehensive

jrahaim opened the pull request.

jrahaim committed:
Support Objects overhaul. List should be comprehensoive

larsacus commented on TeamSnapSDK/SDK/DataLayer/TSDKObjectsRequest.m:
```diff
@@ -73,35 +73,84 @@
-            free(classList);
-        }
-        supportedSDKObjects = [supportedObjects copy];
+        _supportedSDKObjects = @[[TSDKTeamFee class],
```
What is this list used for and when should we be updating it from now on?

jrahaim replied:
this list is used when getting an array of Collection+JSON objects from the server, and converting them into ObjC classes. 

In the instructors for adding classes to the SDK there is a comment you need to import the file into the header for this file. That caused the list to be generated in real time. For the right now, you will need to also add the type to this list.

If this reduces the crashes we should start generating the list at the CI level.

larsacus replied:
If that's the intention, having this list will not make a difference in the contents of the list. `objc_copyClassList` will give you a comprehensive list of all classes loaded into the runtime regardless of if you have imported the header into this class or not. We can import only what we need to compile this class and it will provide the same classes in `objc_copyClassList`.

jrahaim replied:
Are you saying, we simply don't need to import the headers (that makes sense) but we will still need the list to process the objects.. or that we don't need the list at all? (which I  will have to take a much deeper look at when I am able to concentrate).

jrahaim replied:
Actually @larsacus When I was implementing this originally `objc_copyClassList` only contained the classes that this class was exposed to.. I will re-test as part of the tests I am implementing based on Skylers comment.

larsacus replied:
I'm saying this change to this file might not be required at all. I just executed this to verify and `objc_copyClassList` listed GoogleAds, Lottie, Firebase, underscored, and private Apple classes. Everything that I anticipated.

jrahaim replied:
Feel free to close the PR if it's mot needed. I'll take a deeper look another time.

SkylerSeamans replied:
@larsacus check out `+ (TSDKCollectionObject *)teamSnapObjectFromCollectionJSON:(TSDKCollectionJSON *)collectionJSON` to see usage of this array.

larsacus replied:
ah I see that now. yeah seems pretty foundational.

SkylerSeamans commented:
What if you write a test to compare our static list against the dynamic list to insure we always update it?

jrahaim commented:
@SkylerSeamans  That is an interesting option.. It would only work if the headers were updated for that class but the list was not.

SkylerSeamans commented:
I think the test will work even if the header isn't updated. objc_copyClassList(&classCount) should return every class in the runtime.

jrahaim committed:
Add tests for SupportedSDKObjects

jrahaim committed:
Test cleanup

jrahaim commented:
Added test for the new static supportedSDKObjects

SkylerSeamans commented:
Tests look good. 👍 for merge from me!

larsacus commented:
I'm still not convinced we need this change if the runtime has all of the information we need without having to update it.

Unless I'm missing something?

SkylerSeamans commented:
@larsacus https://fabric.io/teamsnap/ios/apps/com.teamsnap.ios/issues/5af0bc7c11e9fa0aa5c9c8cc?time=last-seven-days

SkylerSeamans commented:
This change it to mitigate this crash...which admittedly makes no sense, but I assume was broken with Xcode 9.3.

larsacus commented:
Right I've seen the stack trace and knew this change was to mitigate it, but I still don't understand how this fixes the issue.

SkylerSeamans commented:
It's definitely an experiment. I would think the reduced processing time would at least mask whatever shenanigans are occurring. Open to PRs that address the root cause, if we can find one.

larsacus commented on TeamSnapSDK/SDK/DataLayer/TSDKObjectsRequest.m:
```diff
@@ -73,35 +73,106 @@
+    unsigned int classCount = 0;
+    Class *classList = objc_copyClassList(&classCount);
+
+    for (unsigned int i = 0; i < classCount; i++) {
```
Let's upgrade this block of code to handle the possible exceptions that can occur with `class_getName`:

``` objc
for (unsigned int i = 0; i < classCount; i++) {
    const char* potentialClassName = class_getName(classList[i]);
    if (potentialClassName != NULL &&
        strcmp(potentialClassName, "nil") != 0 &&
        strcmp(potentialClassName, "") != 0) {
        NSString *className = [NSString stringWithUTF8String:potentialClassName];
        if([className hasPrefix:@"TSDK"]) {
            Class class = NSClassFromString(className);
            
            if( [class isEqual:[TSDKCollectionObject class]] == false &&
               [class isEqual:[TSDKShortPropertyTest class]] == false) {
                if(class && [class isSubclassOfClass:[TSDKCollectionObject class]]) {
                    [supportedObjects addObject:class];
                }
            }
        }
    }
}
```

larsacus replied:
Then we can remove line 30/31 in the test for the static list

larsacus replied:
Resolved

larsacus committed:
Adding more robust dynamic class list fetching:
Accounts for internal exceptions related to parsing
mangled swift names from class_getName

larsacus commented:
Added changes to dynamic class list fetch. Test executed successfully.

My entire test suite was seeing random test failures with the `testRequestObjectsForPath` test, but that appears to be unrelated to this change.
